### PR TITLE
bug: Remove CURL* handle from multi-handle on errors.

### PR DESCRIPTION
**Pull request by coryan**
Otherwise the CURL library is left in an inconsistent state. This is part
of the work for #2655.

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/2688)

### CODE REVIEW

**Comment by coryan** [comment]
Looks like this needs more work.  Removing the reviewers for now.

**Comment by coryan** [comment]
Okay, this is working now. PTAL.

File: google/cloud/storage/tests/slow_reader_integration_test.cc [hunk 1]
```diff
@@ -118,6 +118,75 @@ TEST_F(SlowReaderIntegrationTest, StreamingRead) {
...
+TEST_F(SlowReaderIntegrationTest, StreamingReadRestart) {
+  StatusOr<Client> client = Client::CreateDefaultClient();
+  ASSERT_STATUS_OK(client);
+
+  std::string const bucket_name = flag_bucket_name;
+  auto object_name = MakeRandomObjectName();
+  auto file_name = MakeRandomObjectName();
+
+  // Construct a large object, or at least large enough that it is not
+  // downloaded in the first chunk.
```
Comment (devbww) [inline google/cloud/storage/tests/slow_reader_integration_test.cc]: Perhaps just "Construct an object large enough to not be downloaded in the first chunk"?

File: google/cloud/storage/tests/slow_reader_integration_test.cc [hunk 2]
```diff
@@ -118,6 +118,75 @@ TEST_F(SlowReaderIntegrationTest, StreamingRead) {
...
+  auto large_text = CreateLargeText(16 * 1024 * 1024);
+
+  // Create an object with the contents to download.
+  StatusOr<ObjectMetadata> source_meta = client->InsertObject(
+      bucket_name, object_name, large_text, IfGenerationMatch(0));
+  ASSERT_STATUS_OK(source_meta);
+
+  // Create an iostream to read the object back, when running against the
```
Comment (devbww) [inline google/cloud/storage/tests/slow_reader_integration_test.cc]: s/, when/. When/?

Comment (coryan) [reply google/cloud/storage/tests/slow_reader_integration_test.cc]: Fixed.

File: google/cloud/storage/tests/slow_reader_integration_test.cc (hunk 1 above)
Comment (coryan) [reply google/cloud/storage/tests/slow_reader_integration_test.cc]: Fixed.

**Pull request merged**